src/main/java/com/googleinterns/zoomtube/transcriptParser/TranscriptParser.java

```diff
@@ -123,22 +125,56 @@ private Document fetchUrlAsXmlDocument(URL url) throws IOException {
-      String lineContent = StringEscapeUtils.unescapeXml(transcriptNode.getTextContent());
-
-      float lineStartSeconds = Float.parseFloat(transcriptElement.getAttribute(ATTR_START));
-      float lineDurationSeconds = Float.parseFloat(transcriptElement.getAttribute(ATTR_DURATION));
-      // I couldn't find any official way to convert a float seconds to long milliseconds without
-      // losing precision.
-      long lineStartMs = Math.round(lineStartSeconds * MILLISECONDS_PER_SECOND);
-      long lineDurationMs = Math.round(lineDurationSeconds * MILLISECONDS_PER_SECOND);
-      long lineEndMs = lineStartMs + lineDurationMs;
-
-      datastore.put(TranscriptLineUtil.createEntity(
-          lectureKey, lineContent, lineStartMs, lineDurationMs, lineEndMs));
+    Transaction transaction = datastore.beginTransaction();
+
+    try {
+      NodeList transcriptNodes = document.getElementsByTagName(TAG_TEXT);
+      for (int nodeIndex = 0; nodeIndex < transcriptNodes.getLength(); nodeIndex++) {
+        Element transcriptElement = (Element) transcriptNodes.item(nodeIndex);
+        Entity transcriptLineEntity =
+            createTranscriptLineFromElement(lectureKey, transcriptElement);
+        datastore.put(transaction, transcriptLineEntity);
+      }
+
+      transaction.commit();
+    } finally {
+      // If the transaction was interruped, we should make sure it is rolled back
+      // to avoid a partial datastore commit.
+      if (transaction.isActive()) {
+        transaction.rollback();
+      }
     }
   }
+
+  /**
+   * Creates a Transcript Line entity from the XML {@code transcriptLineElement} as part of the
+   * transcript for the lecture referenced by {@code lectureKey}.
+   */
+  private Entity createTranscriptLineFromElement(Key lectureKey, Element transcriptLineElement) {
+    String lineContent = cleanupTranscriptContent(transcriptLineElement.getTextContent());
+
+    float lineStartSeconds = Float.parseFloat(transcriptLineElement.getAttribute(ATTR_START));
+    float lineDurationSeconds = Float.parseFloat(transcriptLineElement.getAttribute(ATTR_DURATION));
+    // I couldn't find any official way to convert a float seconds to long milliseconds without
+    // losing precision.
+    long lineStartMs = Math.round(lineStartSeconds * MILLISECONDS_PER_SECOND);
+    long lineDurationMs = Math.round(lineDurationSeconds * MILLISECONDS_PER_SECOND);
+    long lineEndMs = lineStartMs + lineDurationMs;
+
+    return TranscriptLineUtil.createEntity(
+        lectureKey, lineContent, lineStartMs, lineDurationMs, lineEndMs);
+  }
+
+  /**
+   * Cleans the {@code content} of a transcript line by unescaping XML characters, and
+   * removing newlines.
+   */
+  private String cleanupTranscriptContent(String content) {
+    String unescapedXml = StringEscapeUtils.unescapeXml(content);
+
+    // We ignore '\r' for now.
+    String cleanedContent = unescapedXml.replace('\n', ' ');
+
+    return cleanedContent;
+  }
```

Comment:
can a unit test be added to ensure that it is working?

Reply:
I can using `@VisibleForTesting`, is that a good idea?

Reply:
Ah I guess since it's a bit tricky to test, not having a unit test should be okay. We probably shouldn't test it directly with @visibleForTesting since private methods don't have to be tested.

Reply:
Can't you load a transcript which has \n and then you obtain it and see there are no newlines? 

If you cannot easily test this then its a code smell that the class is doing too much.

Reply:
@amuamushu  private methods should be tested, but through the public methods which call them.

Reply:
Done, although it's rather inelegant.  The only way to test this is with a video id, that results in an API call.  And the only video I know with newlines has a lot of transcript lines, so it increased the total test running time by half a second or so.